File: src/bindings/gi/lua/lgi/override/GElektra.lua
local lgi = require 'lgi'
local GElektra = lgi.GElektra

local core = require 'lgi.core'

local function __func_hide(klass, old, new)
	klass._method[new] = klass._method[old]
	klass._method[old] = nil
end

-- we need to multiplex certain metamethods as the metatable
-- is shared among *all* classes / instances in lgi
local function __create_pseudomt(klass, funcs)
	local mt = getmetatable(klass)
	for _, func in pairs(funcs) do
		local orig_func = mt[func]
		mt[func] = function(self, ...)
			-- using rawget, as we also might wrap __index
			local _type = core.object.query(self, 'repo')
			local _mt = rawget(_type, "__mt")
			if _mt then
				local _hook = rawget(_mt, func)
				if _hook then
					return _hook(self, orig_func, ...)
				end
			end

			return (orig_func) and orig_func(self, ...)
				or error("attempt to call undefined metamethod")
		end
	end
end

-- make the enums global
for n in pairs(GElektra.KeySwitch) do
	if string.upper(n) == n then
		GElektra._enum['KEY_' .. n] = GElektra.KeySwitch[n]
	end
end

for n in pairs(GElektra.KdbOptions) do
	if string.upper(n) == n then
		GElektra._enum['KDB_O_' .. n] = GElektra.KdbOptions[n]
	end
end

GElektra.KS_END = nil

-- setup shared instance hooks
__create_pseudomt(GElektra.Key(), {
		"__eq", "__lt", "__le", "__tostring",
		"__len", "__ipairs",
		--"__index" --disabled, too slow and error-prone
	})

-- Key
-- constructor
__func_hide(GElektra.Key, 'gi_init', '_init')
__func_hide(GElektra.Key, 'gi_make', '_make')
function GElektra.Key:_new(arg0, ...)
	if GElektra.Key:is_type_of(arg0) then
		return GElektra.Key._method._make(arg0)
	end

	local key = GElektra.Key.new(self)
	if not arg0 then
		return key
	end

	local flags = 0
	local value = nil
	local meta  = {}
	local t = { ... }
	local i, arg = next(t, nil)
	while i do
		if arg == GElektra.KEY_END then
			break
		elseif arg == GElektra.KEY_SIZE then
			-- ignore value
			next(t, i)
		elseif arg == GElektra.KEY_VALUE then
			i, value = next(t, i)
		elseif arg == GElektra.KEY_FUNC then
			error("Unsupported meta type")
		elseif arg == GElektra.KEY_FLAGS then
			i, flags = next(t, i)
		elseif arg == GElektra.KEY_META then
			i, k = next(t, i)
			i, meta[k] = next(t, i)
		elseif type(arg) == "number" then
			io.stderr:write("Deprecated option in keyNew: ", arg, "\n")
			flags = bit32.bor(flags, arg)
		elseif GElektra.DEBUG > 0 then
			io.stderr:write("Unknown option in keyNew: ", arg, "\n")
		end
		i, arg = next(t, i)
	end

	-- _init clears our key
	GElektra.Key._method._init(key, arg0, flags, value, value)
	for k, v in pairs(meta) do
		key:setmeta(k, v)
	end
	return key
end

-- operators
__func_hide(GElektra.Key, 'cmp', '_cmp')
GElektra.Key.__mt = GElektra.Key.__mt or {}
function GElektra.Key.__mt:__eq(_, other)
	return GElektra.Key._method._cmp(self, other) == 0
end
function GElektra.Key.__mt:__lt(_, other)
	return GElektra.Key._method._cmp(self, other) < 0
end
function GElektra.Key.__mt:__le(_, other)
	return GElektra.Key._method._cmp(self, other) <= 0
end

function GElektra.Key.__mt:__tostring()
	return self.name
end

-- hide name manipulation functions. use attributes below
__func_hide(GElektra.Key, 'setname',         '_setname')
__func_hide(GElektra.Key, 'setbasename',     '_setbasename')
__func_hide(GElektra.Key, 'getnamesize',     '_getnamesize')
__func_hide(GElektra.Key, 'getbasenamesize', '_getbasenamesize')
__func_hide(GElektra.Key, 'getfullnamesize', '_getfullnamesize')
local keyinvalidname = "Invalid Keyname: keyname needs to start "
		.. "with user/ or system/";
function GElektra.Key:_setname(name)
	ret = GElektra.Key._method._setname(self, name)
	return (ret > 0) and ret or error(keyinvalidname)
end

function GElektra.Key:_setbasename(name)
	ret = GElektra.Key._method._setbasename(self, name)
	return (ret > 0) and ret or error(keyinvalidname)
end

function GElektra.Key:addbasename(name)
	ret = GElektra.Key._method.addbasename(self, name)
	return (ret > 0) and ret or error(keyinvalidname)
end

-- properties / attributes
GElektra.Key._attribute = {}
GElektra.Key._attribute.name = {
	get = function(self)
		return GElektra.Key:_access_property(self, self._property.name)
	end,
	set = GElektra.Key._setname
}
GElektra.Key._attribute.basename = {
	get = function(self)
		return GElektra.Key:_access_property(self, self._property.basename)
	end,
	set = GElektra.Key._setbasename
}
GElektra.Key._attribute.fullname = {
	get = function(self)
		return GElektra.Key:_access_property(self, self._property.fullname)
	end
}

-- hide value manipulation functions. use attributes below
__func_hide(GElektra.Key, 'gi_getstring', '_getstring')
__func_hide(GElektra.Key, 'gi_getbinary', '_getbinary')
__func_hide(GElektra.Key, 'setstring',    '_setstring')
__func_hide(GElektra.Key, 'setbinary',    '_setbinary')
__func_hide(GElektra.Key, 'getvaluesize', '_getvaluesize')
GElektra.Key._attribute.string = {
	get = GElektra.Key._method._getstring,
	set = GElektra.Key._method._setstring
}
GElektra.Key._attribute.binary = {
	get = function(self)
		return tostring(GElektra.Key._method._getbinary(self))
	end,
	set = GElektra.Key._method._setbinary
}
GElektra.Key._attribute.value = {
	get = function(self)
		return (self:isbinary()) and self.binary or self.string
	end
}
GElektra.Key.get = GElektra.Key._attribute.value.get

-- metadata functions
__func_hide(GElektra.Key, 'rewindmeta',  '_rewindmeta')
__func_hide(GElektra.Key, 'nextmeta',    '_nextmeta')
__func_hide(GElektra.Key, 'currentmeta', '_currentmeta')
function GElektra.Key:getmeta(...)
	if select("#", ...) > 0 then
		local meta = GElektra.Key._method.getmeta(self, ...)
		return meta and not meta:isnull() and meta or nil
	end
	return self:__metaiter()
end

function GElektra.Key:__metaiter()
	return coroutine.wrap(
		function()
			GElektra.Key._method._rewindmeta(self)
			local meta = GElektra.Key._method._nextmeta(self)
			while meta and not meta:isnull() do
				coroutine.yield(meta)
				meta = GElektra.Key._method._nextmeta(self)
			end
		end
	)
end

-- KeySet
-- constructor
function GElektra.KeySet:_new(arg0, ...)
	if GElektra.KeySet:is_type_of(arg0) then
		return GElektra.KeySet.dup(arg0)
	end

	local ks = GElektra.KeySet.new(self)
	if arg0 then
		GElektra.KeySet.resize(ks, arg0)
	end

	-- there's no need to check for KS_END
	-- ipairs will do this for us
	for _, arg in ipairs({...}) do
		ks:append(arg)
	end

	return ks
end

-- operators
__func_hide(GElektra.KeySet, 'len', '_len')
GElektra.KeySet.__mt = GElektra.KeySet.__mt or {}
function GElektra.KeySet.__mt:__len()
	return GElektra.KeySet._method._len(self)
end

function GElektra.KeySet.__mt:__index(index_func, prop)
	local key = index_func(self, "lookup")(self, prop)
	return key or index_func(self, prop)
end

__func_hide(GElektra.KeySet, 'rewind',   '_rewind')
__func_hide(GElektra.KeySet, 'next',     '_next')
__func_hide(GElektra.KeySet, 'current',  '_current')
__func_hide(GElektra.KeySet, 'atcursor', '_atcursor')
function GElektra.KeySet.__mt:__ipairs(x, y, z)
	local function ks_ipairs(ks, i)
		local key = ks:lookup(i)
		if key then
			i = i + 1
			return i, key
		end
		return nil
	end
	return ks_ipairs, self, 0
end

__func_hide(GElektra.KeySet, 'gi_append',        'append')
__func_hide(GElektra.KeySet, 'gi_append_keyset', '_append_keyset')
function GElektra.KeySet:append(data)
	if GElektra.KeySet:is_type_of(data) then
		return GElektra.KeySet._method._append_keyset(self, data)
	end
	return GElektra.KeySet._method.append(self, data)
end

__func_hide(GElektra.KeySet, 'lookup_byname', '_lookup_byname')
function GElektra.KeySet:lookup(name)
	local key = nil
	if GElektra.Key:is_type_of(name) then
		key = GElektra.KeySet._method.lookup(self, name, GElektra.KDB_O_NONE)
	elseif type(name) == "string" then
		key = GElektra.KeySet._method._lookup_byname(self, name,
			GElektra.KDB_O_NONE)
	elseif type(name) == "number" then
		key = GElektra.KeySet._method._atcursor(self, name)
	else
		error("Invalid argument type")
	end
	return key
end

-- Kdb
-- constructor
__func_hide(GElektra.Kdb, 'gi_open', 'open')
function GElektra.Kdb:_new(arg0)
	local kdb = GElektra.Kdb.new(self)
	GElektra.Kdb._method.open(kdb, arg0 or GElektra.Key())
	return kdb
end

-- basic methods
function GElektra.Kdb:get(ks, parent)
	if type(parent) == "string" then
		parent = GElektra.Key(parent)
	end
	return GElektra.Kdb._method.get(self, ks, parent)
end

function GElektra.Kdb:set(ks, parent)
	if type(parent) == "string" then
		parent = GElektra.Key(parent)
	end
	return GElektra.Kdb._method.set(self, ks, parent)
end

GElektra.KDB = GElektra.Kdb

File: src/bindings/gi/python/gi/overrides/GElektra.py
from ..module import get_introspection_module
from ..overrides import override
import warnings

GElektra = get_introspection_module('GElektra')

def __func_alias(klass, old, new):
	func = getattr(klass, old)
	setattr(klass, new, func)

def __func_rename(klass, old, new):
	__func_alias(klass, old, new)
	delattr(klass, old)

__all__ = []

## make the enums global
for n in GElektra.KeySwitch.__dict__:
	if n.isupper():
		globals()['KEY_' + n] = getattr(GElektra.KeySwitch, n)
		__all__.append('KEY_' + n)

for n in GElektra.KdbOptions.__dict__:
	if n.isupper():
		globals()['KDB_O_' + n] = getattr(GElektra.KdbOptions, n)
		__all__.append('KDB_O_' + n)

KS_END = None
__all__.append('KS_END')

# exceptions
class Exception(Exception):
	def __init__(self, args = "Exception thrown by Elektra"):
		super().__init__(args)
class KeyException(Exception):
	def __init__(self, args = "Exception thrown by a Key, typically "
			"because you called a method on a null key. "
			"Make sure to check this with !key first"):
		super().__init__(args)
class KeyInvalidName(KeyException):
	def __init__(self, args = "Invalid Keyname: keyname needs to start "
			"with user/ or system/"):
		super().__init__(args)
__all__.extend([ 'Exception', 'KeyException', 'KeyInvalidName' ])

## Key
# rename gi-specific functions
__func_rename(GElektra.Key, 'gi_init',      '_init')
__func_rename(GElektra.Key, 'gi_make',      '_make')
__func_rename(GElektra.Key, 'gi_getstring', '_getstring')
__func_rename(GElektra.Key, 'gi_getbinary', '_getbinary')

# Python API convenience
__func_rename(GElektra.Key, 'cmp', '__cmp__')

__func_rename(GElektra.Key, 'setname',         '_setname')
__func_rename(GElektra.Key, 'setbasename',     '_setbasename')
__func_rename(GElektra.Key, 'getnamesize',     '_getnamesize')
__func_rename(GElektra.Key, 'getbasenamesize', '_getbasenamesize')
__func_rename(GElektra.Key, 'getfullnamesize', '_getfullnamesize')

__func_rename(GElektra.Key, 'setstring',    '_setstring')
__func_rename(GElektra.Key, 'setbinary',    '_setbinary')
__func_rename(GElektra.Key, 'getvaluesize', '_getvaluesize')

__func_rename(GElektra.Key, 'rewindmeta',  '_rewindmeta')
__func_rename(GElektra.Key, 'nextmeta',    '_nextmeta')
__func_rename(GElektra.Key, 'currentmeta', '_currentmeta')

class Key(GElektra.Key):
	def __new__(cls, *args):
		# copy constructor
		if len(args) == 1 and isinstance(args[0], cls):
			return super()._make(args[0])
		return super().__new__(cls, args)

	def __init__(self, *args):
		super().__init__()
		if len(args) == 0:
			return

		arg0, *args = args
		# copy constructor has been used, no init needed
		if isinstance(arg0, self.__class__):
			return

		flags = 0
		value = None
		meta  = {}
		args  = iter(args)
		for arg in args:
			if arg == KEY_END:
				break
			elif arg == KEY_SIZE:
				# ignore value
				next(args)
			elif arg == KEY_VALUE:
				value = next(args)
			elif arg == KEY_FUNC:
				raise TypeError("Unsupported meta type")
			elif arg == KEY_FLAGS:
				flags = next(args)
			elif arg == KEY_META:
				k = next(args)
				meta[k] = next(args)
			elif isinstance(arg, GElektra.KeySwitch):
				warnings.warn("Deprecated option in keyNew: {0}".format(arg),
					DeprecationWarning)
				flags |= arg
			else:
				warnings.warn("Unknown option in keyNew: {0}".format(arg),
					RuntimeWarning)

		# _init clears our key
		if isinstance(value, bytes):
			super()._init(arg0, flags | KEY_BINARY, None, value)
		else:
			super()._init(arg0, flags & ~KEY_BINARY, value, None)
		for k in meta:
			self.setmeta(k, meta[k])

	def _setname(self, name):
		ret = super()._setname(name)
		if ret < 0:
			raise KeyInvalidName()
		return ret

	def _setbasename(self, name):
		ret = super()._setbasename(name)
		if ret < 0:
			raise KeyInvalidName()
		return ret

	def addbasename(self, name):
		ret = super().addbasename(name)
		if ret < 0:
			raise KeyInvalidName()
		return ret

	def get(self):
		"""returns the keys value"""
		if self.isbinary():
			return self._getbinary()
		return self._getstring()

	def set(self, value):
		"""set the keys value. Can be either string or binary"""
		if isinstance(value, bytes):
			return self._setbinary(value)
		return self._setstring(str(value))

	def getmeta(self, name = None):
		"""returns a metakey given by name. Name can be either string or Key.
		If no metakey is found None is returned.
		If name is omitted an iterator object is returned.
		"""
		if name is not None:
			meta = super().getmeta(name)
			return meta if meta else None
		return self.__metaIter()

	def setmeta(self, name, value):
		"""set a new metakey consisting of name and value"""
		if isinstance(value, str):
			return super().setmeta(name, value)
		raise TypeError("Unsupported value type")

	def __metaIter(self):
		self._rewindmeta()
		meta = self._nextmeta()
		while meta:
			yield meta
			meta = self._nextmeta()

	def __str__(self):
		return self.name

	def __bool__(self):
		return not self.isnull();

	def __eq__(self, o):
		return self.__cmp__(o) == 0
	def __ne__(self, o):
		return self.__cmp__(o) != 0

	def __gt__(self, o):
		return self.__cmp__(o) > 0
	def __ge__(self, o):
		return self.__cmp__(o) >= 0

	def __lt__(self, o):
		return self.__cmp__(o) < 0
	def __le__(self, o):
		return self.__cmp__(o) <= 0

	name     = property(lambda self: self.get_property('name'), _setname)
	value    = property(get, set, None, "Key value")
	basename = property(lambda self: self.get_property('basename'), _setbasename)
	fullname = property(lambda self: self.get_property('fullname'))

Key = override(Key)
__all__.append('Key')

## KeySet
# rename gi-specific functions
__func_rename(GElektra.KeySet, 'gi_append',        'append')
__func_rename(GElektra.KeySet, 'gi_append_keyset', '_append_keyset')

# Python API convenience
__func_rename(GElektra.KeySet, 'len', '__len__')

__func_rename(GElektra.KeySet, 'lookup_byname', '_lookup_byname')

__func_rename(GElektra.KeySet, 'rewind',   '_rewind')
__func_rename(GElektra.KeySet, 'next',     '_next')
__func_rename(GElektra.KeySet, 'current',  '_current')
__func_rename(GElektra.KeySet, 'atcursor', '_atcursor')

class KeySet(GElektra.KeySet):
	def __new__(cls, *args):
		if len(args) == 1 and isinstance(args[0], __class__):
			return super().dup(args[0])
		return super().__new__(cls, args)

	def __init__(self, *args):
		super().__init__()
		if len(args) == 0:
			return

		arg0, *args = args
		if isinstance(arg0, __class__):
			return

		self.resize(arg0)
		for arg in args:
			if arg is KS_END:
				break
			self.append(arg)

	def lookup(self, name):
		"""Lookup a key by name. Name can be either string, Key or indexes.
		If index is negative, search starts at the end.
		Returns None if no key is found.
		"""
		if isinstance(name, Key):
			key = super().lookup(name, KDB_O_NONE)
		elif isinstance(name, str):
			key = self._lookup_byname(name, KDB_O_NONE)
		elif isinstance(name, int):
			key = self._atcursor(name)
		else:
			raise TypeError("Unsupported type")
		return key if key else None

	def append(self, data):
		if isinstance(data, __class__):
			return self._append_keyset(data)
		return super().append(data)

	def __getitem__(self, key):
		"""See lookup(...) for details.
		Slices and negative indexes are supported as well.
		"""
		if isinstance(key, slice):
			return [ self[k] for k in range(*key.indices(len(self))) ]
		elif isinstance(key, ( int )):
			item = self.lookup(key)
			if item is None:
				raise IndexError("index out of range")
			return item
		elif isinstance(key, ( str, Key )):
			item = self.lookup(key)
			if item is None:
				raise KeyError(str(key))
			return item
		raise TypeError("Invalid argument type")

	def __contains__(self, item):
		"""See lookup(...) for details"""
		if isinstance(item, ( str, Key )):
			key = self.lookup(item)
			return True if key else False
		raise TypeError("Invalid argument type")

	def __iter__(self):
		i = 0
		key = self.lookup(i)
		while key:
			yield key
			i = i + 1
			key = self.lookup(i)

KeySet = override(KeySet)
__all__.append('KeySet')

## Kdb
# rename gi-specific functions
__func_rename(GElektra.Kdb, 'gi_open', 'open')

class Kdb(GElektra.Kdb):
	def __init__(self, *args):
		super().__init__()
		self.open(args[0] if len(args) else Key())

	def get(self, ks, parent):
		if isinstance(parent, str):
			parent = Key(parent)
		return super().get(ks, parent)

	def set(self, ks, parent):
		if isinstance(parent, str):
			parent = Key(parent)
		super().set(ks, parent)

	def __enter__(self):
		"""Internal method for usage with context managers"""
		return self

	def __exit__(self, type, value, tb):
		"""Internal method for usage with context managers.
		Closes the database.
		"""
		try:
			self.close(Key())
		except:
			pass

Kdb = override(Kdb)
KDB = Kdb
__all__.extend([ 'Kdb', 'KDB' ])
